app/src/main/java/com/bnyro/contacts/services/CallService.kt

```diff
@@ -50,32 +61,93 @@ class CallService : InCallService() {
         call.registerCallback(callCallback)
         CallManager.setCall(call)
 
-        val notification = buildNotification(call)
-        NotificationManagerCompat.from(this).notify(CALL_NOTIFICATION_ID, notification)
+        scope.launch {
+            val callerNumber = CallManager.callerDisplayNumber
+            val (thumbnailUri, contactName) = withContext(Dispatchers.IO) {
+                getContactName(callerNumber)
+            }
+            val contactPhoto = if (thumbnailUri != null) {
+                withContext(Dispatchers.IO) {
+                    getContactPhotoThumbnail(thumbnailUri)
+                }
+            } else {
+                null
+            }
+
+            val notification = buildNotification(call, callerNumber, contactName, contactPhoto)
+            NotificationManagerCompat.from(this@CallService)
+                .notify(CALL_NOTIFICATION_ID, notification)
 
+        }
         val intent = Intent(applicationContext, CallActivity::class.java)
         intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
         startActivity(intent)
     }
 
+    private fun getContactName(phoneNumber: String): Pair<String?, String?> {
+        val cr: ContentResolver = this.contentResolver
+        val uri = Uri.withAppendedPath(PhoneLookup.CONTENT_FILTER_URI, Uri.encode(phoneNumber))
+        val query =
+            cr.query(
+                uri,
+                arrayOf(PhoneLookup.DISPLAY_NAME, PhoneLookup.PHOTO_THUMBNAIL_URI),
+                null,
+                null,
+                null
+            )
+                ?: return null to null
+        var contactName: String? = null
+        var contactThumbnail: String? = null
+
+        query.use { cursor ->
+            if (cursor.moveToFirst()) {
+                val nameIndex = cursor.getColumnIndex(PhoneLookup.DISPLAY_NAME)
+                val thumbnailIndex = cursor.getColumnIndex(PhoneLookup.PHOTO_THUMBNAIL_URI)
+                contactName = cursor.getString(nameIndex)
+                contactThumbnail = cursor.getString(thumbnailIndex)
+            }
+        }
+
+        return contactThumbnail to contactName
+    }
+
+    private fun getContactPhotoThumbnail(uri: String): Bitmap? {
+        val contentResolver = this.contentResolver
+        val photoThumbnailUri = Uri.parse(uri)
+        val assetFileDescriptor =
+            contentResolver.openAssetFileDescriptor(photoThumbnailUri, "r") ?: return null
+        val fileDescriptor = assetFileDescriptor.fileDescriptor
+        val bitmap = BitmapFactory.decodeFileDescriptor(fileDescriptor)
+        assetFileDescriptor.close()
+        return bitmap
```

Comment:
We probably already do almost the same in the `DeviceContactsRepository`, so we should move this tothe `ContactHelper` object or so.